Introduces `optuna._imports.try_import` to DRY optional imports.

hvy opened the pull request:
## Motivation

DRY. Optional imports are currently associated with code duplication.

## Description of the changes

Introduces `optuna._imports` and `optuna._imports.try_import` to DRY optional imports. Note that deferred error messages are somewhat simplified/generalized.

## Note

Other components such as `optuna._LazyImport` and `optuna.integration._IntegrationModule` could probably be moved into the introduced `optuna._imports` in follow up work.

toshihikoyanase commented on optuna/_imports.py:
```diff
@@ -0,0 +1,97 @@
+        self,
+        exc_type: Optional[Type[Exception]],
+        exc_value: Optional[Exception],
+        traceback: Optional[TracebackType],
```
Currently, the `traceback` argument is not used. This is not a strong opinion, but how about showing it with the error message?
IMO, it helps users to find the issue of the code.
For example, the current message shown in the console is as follows:

```console
>>> optuna.integration.ChainerPruningExtension(None, None, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/optuna/optuna/integration/chainer.py", line 53, in __init__
    _imports.check()
  File "/tmp/optuna/optuna/_imports.py", line 88, in check
    raise exc_type(msg)
ImportError: Failed to import an optional package. No module named 'chainer'.
```

It does not contain the line where `chainer` was imported, and users need to guess it.
On the other hand, `traceback` has the line as follows:
```console
  File "/tmp/optuna/optuna/integration/chainer.py", line 5, in <module>
    import chainer
```

hvy replied:
Thanks, that makes sense. We can probably nest the exceptions using `raise ... from ...`. I'll push the change. I think it'll include enough information to satisfy https://github.com/optuna/optuna/pull/1315#pullrequestreview-423254751 although I agree that the former was friendlier.

hvy replied:
Example after https://github.com/optuna/optuna/pull/1315/commits/43a31f67e6af930f94abe9744e4e5a52c2c1ff86

```
Traceback (most recent call last):
  File "/Users/hvy/repos/optuna/optuna/integration/chainer.py", line 5, in <module>
    import chainer
ModuleNotFoundError: No module named 'chainer'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/hvy/repos/optuna/optuna/study.py", line 732, in _run_trial
    result = func(trial)
  File "test_try_import.py", line 6, in objective
    ChainerPruningExtension(trial, "validation/main/accuracy", (1, "epoch"))
  File "/Users/hvy/repos/optuna/optuna/integration/chainer.py", line 53, in __init__
    _imports.check()
  File "/Users/hvy/repos/optuna/optuna/_imports.py", line 80, in check
    raise exc_type(self._message) from exc_value
ModuleNotFoundError: Failed to import an optional package. See causing exceptions for details.
Traceback (most recent call last):
  File "/Users/hvy/repos/optuna/optuna/integration/chainer.py", line 5, in <module>
    import chainer
ModuleNotFoundError: No module named 'chainer'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test_try_import.py", line 12, in <module>
    study.optimize(objective, n_trials=10)
  File "/Users/hvy/repos/optuna/optuna/study.py", line 334, in optimize
    func, n_trials, timeout, catch, callbacks, gc_after_trial, None
  File "/Users/hvy/repos/optuna/optuna/study.py", line 677, in _optimize_sequential
    self._run_trial_and_callbacks(func, catch, callbacks, gc_after_trial)
  File "/Users/hvy/repos/optuna/optuna/study.py", line 708, in _run_trial_and_callbacks
    trial = self._run_trial(func, catch, gc_after_trial)
  File "/Users/hvy/repos/optuna/optuna/study.py", line 732, in _run_trial
    result = func(trial)
  File "test_try_import.py", line 6, in objective
    ChainerPruningExtension(trial, "validation/main/accuracy", (1, "epoch"))
  File "/Users/hvy/repos/optuna/optuna/integration/chainer.py", line 53, in __init__
    _imports.check()
  File "/Users/hvy/repos/optuna/optuna/_imports.py", line 80, in check
    raise exc_type(self._message) from exc_value
ModuleNotFoundError: Failed to import an optional package. See causing exceptions for details.
```

HideakiImamura replied:
Thanks for the update. My concerns have been addressed by this fix.

toshihikoyanase replied:
Awesome! Thank you for your update.

hvy commented:
Thanks for the quick review. I kept the stack trace in the latest change. PTAL.

hvy commented on tests/test_imports.py:
```diff
@@ -0,0 +1,26 @@
+
+    # Non-default arguments.
+    with try_import(catch=(SyntaxError,)) as imports:
+        raise SyntaxError
```
Thanks, applied!

toshihikoyanase commented on optuna/importance/_fanova.py:
```diff
@@ -121,12 +117,3 @@ def _distribution_to_hyperparameter(name: str, distribution: BaseDistribution) -
-
-def _check_fanova_availability() -> None:
-    if not _available:
-        raise ImportError(
-            "fanova is not available. Please install automl/fanova to use this feature. "
-            "For further information, please refer to the installation guide of automl/fanova. "
-            "(The actual import error is as follows: " + str(_import_error) + ")."
-        )
```
This evaluator will be removed by #1106, and this is an optional comment.
IMO, this information is also useful because `fanova` cannot be installed via PyPI. How about keeping it?

The current message is as follows:
```console
Traceback (most recent call last):
  File "/Users/yanase/pfn/code/optuna/optuna/importance/_fanova.py", line 26, in <module>
    from fanova import fANOVA
ModuleNotFoundError: No module named 'fanova'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/yanase/pfn/code/optuna/optuna/_experimental.py", line 86, in wrapped_init
    _original_init(self, *args, **kwargs)
  File "/Users/yanase/pfn/code/optuna/optuna/importance/_fanova.py", line 48, in __init__
    _imports.check()
  File "/Users/yanase/pfn/code/optuna/optuna/_imports.py", line 80, in check
    raise exc_type(self._message) from exc_value
ModuleNotFoundError: Failed to import an optional package. See causing exceptions for details.
```

hvy replied:
Do you mean not to use `try_import`? `fanova` is available on PyPI (although it requires swig).

toshihikoyanase commented on optuna/integration/allennlp.py:
```diff
@@ -88,7 +84,7 @@ def __init__(
     ):
 
-        _check_allennlp_availability()
+        _imports.check()
```
The error message is shown here is as follows:

```console
>>> optuna.integration.AllenNLPExecutor(None, None, None)
/tmp/optuna/optuna/_experimental.py:83: ExperimentalWarning: AllenNLPExecutor is experimental (supported from v1.4.0). The interface can change in the future.
  ExperimentalWarning,
Traceback (most recent call last):
  File "/tmp/optuna/optuna/integration/allennlp.py", line 16, in <module>
    import _jsonnet
ModuleNotFoundError: No module named '_jsonnet'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/optuna/optuna/_experimental.py", line 86, in wrapped_init
    _original_init(self, *args, **kwargs)
  File "/tmp/optuna/optuna/integration/allennlp.py", line 87, in __init__
    _imports.check()
  File "/tmp/optuna/optuna/_imports.py", line 80, in check
    raise exc_type(self._message) from exc_value
ModuleNotFoundError: Failed to import an optional package. See causing exceptions for details.
```

The import error is about `_jsonnet`, which is one of the dependencies of AllenNLP.
How about specifying the module name with the argument of the `try_import` function?

hvy commented:
Thanks for your careful feedback. Having the caller passing a string is somewhat verbose and the information is duplicated. We can seemingly get the module name from the `ImportError`. Although it might not work for `SyntaxError` it should work for other cases. What do you think of using that to construct the error message? As for `_jsonnnet`, we should conditionally import it once `allennlp` is successful.

HideakiImamura commented:
In the case of `allennlp` or `_jsonnet`, I think the current error message is enough to debug it. The user can find that `allennlp.py` causes `ModuleNotFoundError` from the current error message.

I think https://github.com/optuna/optuna/pull/1315#discussion_r435195083 is reasonable. It is useful to include the name of the library in the error message.

hvy commented:
I tried to keep the usage simple without introducing additional arguments while providing better error messages with a fixed raised error type.

```
>>> optuna.integration.AllenNLPExecutor(None, None, None)
/Users/hvy/repos/optuna/optuna/_experimental.py:78: ExperimentalWarning:

AllenNLPExecutor is experimental (supported from v1.4.0). The interface can change in the future.

Traceback (most recent call last):
  File "/Users/hvy/repos/optuna/optuna/integration/allennlp.py", line 15, in <module>
    import allennlp.commands
ModuleNotFoundError: No module named 'allennlp'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/hvy/repos/optuna/optuna/_experimental.py", line 86, in wrapped_init
    _original_init(self, *args, **kwargs)
  File "/Users/hvy/repos/optuna/optuna/integration/allennlp.py", line 89, in __init__
    _allennlp_imports.check()
  File "/Users/hvy/repos/optuna/optuna/_imports.py", line 89, in check
    raise ImportError(message) from exc_value
ImportError: Tried to import 'allennlp' but failed. Please make sure that the package is installed correctly to use this feature. Actual error: No module named 'allennlp'.
```

```
optuna dashboard --study foo --storage sqlite:///dummy.db
[W 2020-06-05 17:19:28,964] The use of `--study` is deprecated. Please use `--study-name` instead.
[E 2020-06-05 17:19:29,060] Tried to import 'bokeh' but failed. Please make sure that the package is installed correctly to use this feature. Actual error: No module named 'bokeh'.
```

hvy commented:
PTAL.

toshihikoyanase commented on optuna/integration/allennlp.py:
```diff
@@ -36,6 +33,8 @@ def dump_best_config(input_config_file: str, output_config_file: str, study: opt
             Note that :func:`~optuna.study.Study.optimize` must have been called.
 
     """
+    _jsonnet_imports.check()
```
How about checking the availability of allennlp? It's also required, and the error message may be more helpful.

hvy replied:
Sorry I somehow missed that AllenNLP is required here. We probably only need the AllenNLP `_DeferredImportExceptionContextManager` .

hvy replied:
Fixed.

hvy commented:
Thanks @toshihikoyanase as always. Do you think this design looks fine otherwise? I addressed your comment about AllenNLP and also applied this context manager for MDI. 

PTAL.
